README.md

```diff
@@ -29,232 +29,110 @@ If you have used WaterLily for research, please __cite us__! The following manus
-)
-
-# make the writer
-writer = vtkWriter("advanced_writer"; attrib=custom_attrib)
-...
-close(writer)
-```
-The functions that are passed to the `attrib` (custom attributes) must follow the same structure as what is shown in this example, that is, given a `Simulation`, return an N-dimensional (scalar or vector) field. The `vtkwriter` function will automatically write the data to a `.pvd` file, which can be read by ParaView. The prototype for the `vtkwriter` function is:
-```julia
-# prototype vtk writer function
-custom_vtk_function(a::Simulation) = ... |> Array
-```
-where `...` should be replaced with the code that generates the field you want to write to the file. The piping to a (CPU) `Array` is necessary to ensure that the data is written to the CPU before being written to the file for GPU simulations.
+## Multi-threading and GPU backends
 
-### Restarting from a VTK file
+WaterLily uses [KernelAbstractions.jl](https://github.com/JuliaGPU/KernelAbstractions.jl) to multi-thread on CPU and run on GPU backends. The implementation method and speed-up are documented in our [paper](https://physics.paperswithcode.com/paper/waterlily-jl-a-differentiable-and-backend), with costs as low as 1.44 nano-seconds measured per degree of freedom and time step!
 
-The restart of a simulation from a VTK file is demonstrated in [this example](https://github.com/WaterLily-jl/WaterLily.jl/blob/master/examples/ThreeD_cylinder_vtk_restart.jl). The `ReadVTK` package is used to read simulation data from a `.pvd` file. This `.pvd` __must__ have been written with the `vtkwriter` function and __must__ contain at least the `velocity` and `pressure` fields. The following example demonstrates how to restart a simulation from a `.pvd` file using the `ReadVTK` package and the WaterLily `vtkreader` function
-```julia
-using WaterLily,ReadVTK
-sim = make_sim(...)
-# restart the simulation
-writer = restart_sim!(sim; fname="file_restart.pvd")
+Note that multi-threading requires _starting_ Julia with the `--threads` argument, see [the multi-threading section](https://docs.julialang.org/en/v1/manual/multi-threading/) of the manual. If you are running Julia with multiple threads, KernelAbstractions will detect this and multi-thread the loops automatically. 
 
-# append sim data to the file used for restart
-write!(writer, sim)
+Running on a GPU requires initializing the `Simulation` memory on the GPU, and care needs to be taken to move the data back to the CPU for visualization. For example, the cylinder example above could be run `using CUDA` by
+```Julia
+function circle(n,m;Re=100,U=1,T=Float64,mem=Array)
+    radius, center = m/8, m/2-1
+    body = AutoBody((x,t)->√sum(abs2, x .- center) - radius)
+    Simulation((n,m),(U,0),2radius;ν=U*2radius/Re,body,
+                mem, # memory type
+                T)   # Floating point type
+end
 
-# don't forget to close the file
-close(writer)
+using CUDA
+@assert CUDA.functional()   # is your CUDA GPU working?? 
```

Comment:
Is that still useful? The extension does it when you load it already, right?

Reply:
On my machine at home, `using CUDA` doesn't throw any error, so the `@assert` is needed to stop people from following the example blindly.